File: demos_code_J1_conception_objet/02_ocp_strategie/DemoOcpAvant.java
public class DemoOcpAvant {
    enum ModeLivraison { STANDARD, EXPRESS, RELAIS }
    record Commande(int totalCents) {}

    static int frais(Commande c, ModeLivraison mode) {
        return switch (mode) {
            case STANDARD -> c.totalCents() >= 5000 ? 0 : 500;
            case EXPRESS -> 1200;
            case RELAIS -> 300;
        };
    }

    static int totalAvecLivraison(Commande c, ModeLivraison mode) {
        return c.totalCents() + frais(c, mode);
    }

    public static void main(String[] args) {
        Commande c = new Commande(4200);
        for (ModeLivraison mode : ModeLivraison.values()) {
            System.out.println(mode + " -> " + totalAvecLivraison(c, mode));
        }
    }
}

File: Demo en Python/demos_code_J1_conception_objet_python/02_ocp_strategie/DemoOcpAvant.py
from dataclasses import dataclass
from enum import Enum, auto


class ModeLivraison(Enum):
    STANDARD = auto()
    EXPRESS = auto()
    RELAIS = auto()


@dataclass(frozen=True)
class Commande:
    total_cents: int


def frais(c: Commande, mode: ModeLivraison) -> int:
    if mode is ModeLivraison.STANDARD:
        return 0 if c.total_cents >= 5000 else 500
    if mode is ModeLivraison.EXPRESS:
        return 1200
    if mode is ModeLivraison.RELAIS:
        return 300
    raise ValueError(f"Mode de livraison inconnu : {mode}")


def total_avec_livraison(c: Commande, mode: ModeLivraison) -> int:
    return c.total_cents + frais(c, mode)


def main() -> None:
    c = Commande(4200)
    for mode in ModeLivraison:
        print(f"{mode.name} -> {total_avec_livraison(c, mode)}")


if __name__ == "__main__":
    main()
